tests/transports/stream_tests.nim

```diff
@@ -0,0 +1,410 @@
+        except CatchableError:
+          raiseAssert "should not happen"
+        finally:
+          await stream.close()
+      let muxerTask = muxer.handle()
+      asyncSpawn muxerTask
+      await muxerTask
+      await muxer.close()
+      await conn.close()
+
+    proc runClient(server: Transport) {.async.} =
+      let client = transportProvider()
+      let conn = await client.dial("", server.addrs[0])
+      let muxer = muxerProvider(client, conn)
+      let muxerTask = muxer.handle()
+      asyncSpawn muxerTask
+
+      let stream = await muxer.newStream()
+
+      # Client reads server data
+      var buffer: array[serverMessage.len, byte]
+      await stream.readExactly(addr buffer, serverMessage.len)
+      check string.fromBytes(buffer) == serverMessage
+
+      # Server has closed write side, so further reads should EOF
+      if (isQuicTransport(ma[0])):
+        expect LPStreamEOFError:
+          discard await stream.readOnce(addr buffer, 1)
+      else:
+        # TCP/Mplex: First readOnce after closeWrite returns 0
+        let bytesRead = await stream.readOnce(addr buffer, 1)
+        check bytesRead == 0
+
+      # Client should still be able to write back to server
+      await stream.write(clientMessage)
+
+      await stream.close()
+      await muxer.close()
+      await conn.close()
+      await muxerTask
+
+    let server = transportProvider()
+    await server.start(ma)
+    let serverTask = serverHandler(server)
+    asyncSpawn serverTask
+
+    await runClient(server)
+    await serverTask
+    await server.stop()
+
+  asyncTest "stream caching with multiple partial reads":
+    let ma = @[MultiAddress.init(address).tryGet()]
+    const messageSize = 2048
+    const chunkSize = 256
+    let message = (0 ..< messageSize).mapIt(byte(it mod 256))
+
+    proc serverHandler(server: Transport) {.async.} =
+      let conn = await server.accept()
+      let muxer = muxerProvider(server, conn)
+      muxer.streamHandler = proc(stream: Connection) {.async: (raises: []).} =
+        try:
+          await stream.write(message)
+        except CatchableError:
+          discard
+        finally:
+          await stream.close()
+      let muxerTask = muxer.handle()
+      asyncSpawn muxerTask
+      await muxerTask
+      await muxer.close()
+      await conn.close()
+
+    proc runClient(server: Transport) {.async.} =
+      let client = transportProvider()
+      let conn = await client.dial("", server.addrs[0])
+      let muxer = muxerProvider(client, conn)
+      let muxerTask = muxer.handle()
+      asyncSpawn muxerTask
+
+      let stream = await muxer.newStream()
+
+      var receivedData: seq[byte] = @[]
+
+      # Read a message chunk by chunk
+      while receivedData.len < messageSize:
+        var chunk: array[chunkSize, byte]
+        let bytesRead = await stream.readOnce(addr chunk[0], chunkSize)
+        check bytesRead > 0
+        receivedData.add(chunk[0 ..< bytesRead])
```

Comment:
let's have this in utility `readStreamTillEOF`, this utility can be part of new `tools/streams` group.
https://github.com/vacp2p/nim-quic/blob/0d5561fa25aa573ea3465e339330ff030b7bbf7e/tests/helpers/stream.nim#L11